controlplane/kubeadm/controllers/controller.go

```diff
@@ -512,7 +512,7 @@ func (r *KubeadmControlPlaneReconciler) adoptMachines(ctx context.Context, kcp *
 			return err
 		}
 
-		patchHelper, err := patch.NewHelper(m, r.Client)
+		patchHelper, err := patch.NewHelper(m, r.Client, patch.WithStatusObservedGeneration{})
```

Comment:
We should not update observed generation for resources that we do not manage, only the Machine controller should update the observed generation for Machines.
```suggestion
		patchHelper, err := patch.NewHelper(m, r.Client)
```